torch/csrc/autograd/function.h

```diff
@@ -90,9 +139,32 @@ struct Function : std::enable_shared_from_this<Function> {
   };
 
   // Computes is_executable and next_functions from a list of input variables
-  static FunctionFlags flags(const variable_list& inputs);
-  static FunctionFlags flags(const std::initializer_list<Variable>& inputs);
-  static FunctionFlags flags(at::TensorList inputs);
+  // (but which are tensors, because this is what happens in the autogenerated
+  // code)
+  template<typename... Args> inline static FunctionFlags tensor_flags(Args... args) {
+    FunctionFlags f;
+    f.is_executable = false;
+    f.next_functions.resize(countTensors(args...));
+    if (!GradMode::is_enabled()) {
+      // TODO: avoid allocating next_functions entirely if grad_mode is disabled
+      return f;
+    }
+    set_tensor_function_flags(f, 0, args...);
+    return f;
+  }
+
+  // Computes is_executable and next_functions from a list of input variables
+  template<typename... Args> inline static FunctionFlags flags(Args... args) {
+    FunctionFlags f;
+    f.is_executable = false;
+    f.next_functions.resize(countVariables(args...));
```

Comment:
I think it would be quite easy to just have a hard-coded value of e.g. 5 and use `reserve` + `push_back` later. There are very few ops that have more inputs than that.

Reply:
Since `countVariables` has already been written, and `countVariables` is pretty cheap, I can't think of a good reason to not do it the correct way except maybe because you want me to delete `countVariables` and `countTensors`...

Reply:
You guessed it :wink:

Reply:
Eh I'm ok with it staying, it's just that variadic templates are incredibly ugly